java/src/jmri/jmrix/ipocs/protocol/Message.java

```diff
@@ -0,0 +1,85 @@
+package jmri.jmrix.ipocs.protocol;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import jmri.jmrix.ipocs.protocol.packets.Packet;
+
+/**
+ * Represents a IPOCS Message.
+ *
+ * Protocol details can be found on the project website, https://ipocsmr.github.io
+ *
+ * @author Fredrik Elestedt Copyright (C) 2020
+ * @since 4.21.2
+ */
+public class Message {
+  private final static Logger log = LoggerFactory.getLogger(Message.class);
+  private byte length;
+  private String objectName;
+  private List<Packet> packets = new ArrayList<Packet>();
+
+  public byte getLength() {
+    return length;
+  }
+
+  public String getObjectName() {
+    return objectName;
+  }
+
+  public void setObjectName(String objectName) {
+      this.objectName = objectName;
+  }
+
+  public List<Packet> getPackets() {
+    return packets;
+  }
+
+  public static Message parse(ByteBuffer buffer, Integer contentSize) {
+    int startPos = buffer.position();
+    Message msg = new Message();
+    // Get length
+    msg.length = buffer.get();
+    // Ensure that there are enough bytes in the buffer:
+    if (msg.length > contentSize) {
+      return null;
+    }
+    // Get object name
+    StringBuilder sb = new StringBuilder();
+    byte last;
+    while ((last = buffer.get()) != 0x00) {
+      sb.append((char) last);
+    }
+    msg.objectName = sb.toString();
+    // Parse packets
+    Packet packet;
+    log.debug("Message for {}", msg.objectName);
+    while (buffer.position() - startPos < msg.length && (packet = Packet.parse(buffer)) != null) {
+      msg.packets.add(packet);
+    }
+    return msg;
+  }
+
+  public ByteBuffer serialize() {
+    ByteBuffer buffer = ByteBuffer.allocate(1 + objectName.length() + 1);
+    buffer.put(length);
+    buffer.put(objectName.getBytes());
+    buffer.put((byte)0);
+    for (Packet packet : packets) {
+      ByteBuffer serPacket = packet.serialize();
+      ByteBuffer oldBuffer = buffer;
+      oldBuffer.rewind();
+      serPacket.rewind();
+      buffer = ByteBuffer.allocate(buffer.capacity() + serPacket.capacity());
+      buffer.put(oldBuffer);
+      buffer.put(serPacket);
+    }
+    buffer.put(0, (byte)buffer.capacity());
```

Comment:
Should this line put the size of the buffer at byte 0? If so, are you sure that the length of the serialized packets will not exceed 255 bytes?

Reply:
Yes, it's supposed to go at byte 0. It's not until the rest of the serialisation is done that we know the length of the message.
The protocol does not allow for longer messages - so yes :D but you're right - probably should add checks in various places (like checking the length of the beans user names) to make sure it will never happen.

Reply:
Please add a check that the buffer is not greater than 255 bytes. At least write a message to the log:

`if (buffer.capacity() > 255) log.error("Serialized message is longer than 255 bytes");`

That gives us an error message in the log in case it goes wrong for a user. Or you may throw an exception if this happens.

Reply:
Yeah - I will throw an exception there. The only way that will happen is if a userName on a object in the layout is longer than say 230 characters...

Reply:
Thanks!